decode.go

```diff
@@ -471,6 +471,80 @@ func (uttam UnrecognizedTagToAnyMode) valid() bool {
 	return uttam >= 0 && uttam < maxUnrecognizedTagToAny
 }
 
+// SimpleValueRegistry is an immutable mapping from CBOR simple value number (0...23 and 32...255)
+// to Go analog value.
+type SimpleValueRegistry struct {
+	analogs [256]*interface{}
+}
+
+// WithSimpleValueAnalog registers a Go analog value for the given simple value. When decoding into
+// an empty interface value, the registered analog value is returned. When decoding into a concrete
+// type, the type of the registered analog must be directly assignable to the destination's type.
+func WithSimpleValueAnalog(sv SimpleValue, analog interface{}) func(*SimpleValueRegistry) error {
+	return func(r *SimpleValueRegistry) error {
+		if sv >= 24 && sv <= 31 {
+			return fmt.Errorf("cbor: cannot set analog for reserved simple value %d", sv)
+		}
+		r.analogs[sv] = &analog
```

Comment:
I'm not confident that the library should forbid this. As you implied, it's not trivial to pin down without making the limitations very strict.

If limited to only a few built-in types, it becomes hard for users to distinguish between a simple value and a different value with an equivalent Go representation. If the limitation is relaxed to be based on kind rather than type, then users could define a named type with one of the supported underlying types, and use an analog of that type to distinguish the simple value.

However, I think that would eventually bite users who register analogs that they don't entirely control. If a perfect analog value is provided by a library, users may be forced to perform a second pass over their decoded object to convert from a user-defined sentinel value to the analog that they want to appear in the final object. Alternatively,  a library-provided value may change during a dependency update in such a way that it no longer meets our restrictions on acceptable analog values.

I don't have a use case for this presently, but I lean toward leaving users responsible for recognizing that their `*int` analog is the pointer itself and not the referenced value. Can you imagine this causing thorny issues down the line within this library? Map key equivalence is the main area that comes to my mind.